### PR TITLE
Configures the SDK for testing in API coverage environment

### PR DESCRIPTION
**Pull request by dborsatto**
It's mostly changes to the tests, except 2 things:
- `Contentful\Client` now has the method `getLogger` (which maybe could go in the changelog)
- `Contentful\Exception\ApiExeption` now uses `getHeaderLine` for accessing `X-Contentful-Request-Id`, which is safer as it returns an empty string if no header with that name is available.

The rest of the changes are in the `tests` directory, where basically I extracted all `Client` instantiations and moved them to the base `End2EndTestCase` class. As some of the tests involved cleaning up a cache directory, I simply moved these operations to the base `setUp` and `tearDown` methods.

Tests that are not suitable for running in a proxy environment are marked with the PHPUnit annotation `@requires Contentful`, and automatically skipped.

### CODE REVIEW

**Comment by dborsatto** [comment]
I'm not crazy about creating all those instances in `End2EndTestCase::setUpBeforeClass` instead of lazy-loading them, but to be honest most of the times when the test suite is executed, all those client instances are actually needed, so that's probably a reasonable approach.

File: src/Exception/ApiException.php [hunk 1]
```diff
@@ -50,7 +50,7 @@ public function __construct(GuzzleRequestException $previous, $message = '')
 
         if ($this->previous->hasResponse()) {
             $this->response = $this->previous->getResponse();
-            $this->requestId = $this->response->getHeader('X-Contentful-Request-Id')[0];
+            $this->requestId = $this->response->getHeaderLine('X-Contentful-Request-Id');
```
Comment (realityking) [inline src/Exception/ApiException.php]: What's the issue you ran into?

From the PSR-7 spec:
> Use the `getHeaderLine()` method to retrieve a header value as a string containing all header values of a case-insensitive header by name concatenated with a comma.

I'm not sure that's why we want.

File: phpunit-api-coverage.xml [hunk 2]
```diff
@@ -13,4 +11,8 @@
             <directory suffix=".php">src</directory>
         </whitelist>
     </filter>
+
+    <php>
+        <const name="SDK_TESTING_ENV" value="api_coverage" />
```
Comment (realityking) [inline phpunit-api-coverage.xml]: Probably just a question of style but did you consider using an env variable instead?

File: tests/E2E/GzipEncodingTest.php [hunk 3]
```diff
@@ -30,7 +28,7 @@ public function testContentEncodingHeader()
 
         $this->assertEquals('gzip', $logEntry->getRequest()->getHeaderLine('Accept-Encoding'));
 
-        // Need to check 'x-encoded-content-encoding' as curl is automatically decompressing the response
-        $this->assertEquals('gzip', $logEntry->getResponse()->getHeaderLine('x-encoded-content-encoding'));
+        // Need to check 'X-Encoded-Content-Encoding' as curl is automatically decompressing the response
+        $this->assertEquals('gzip', $logEntry->getResponse()->getHeaderLine('X-Encoded-Content-Encoding'));
```
Comment (realityking) [inline tests/E2E/GzipEncodingTest.php]: Is this just a stylistic change? Because either should work.

File: tests/bootstrap-api-coverage.php [hunk 4]
```diff
@@ -1,7 +1,11 @@
 <?php
+
 /**
- * @copyright 2015-2016 Contentful GmbH
+ * This file is part of the contentful.php package.
```
Comment (realityking) [inline tests/bootstrap-api-coverage.php]: Personal preference but I'd not mix this within one project :/

File: tests/End2EndTestCase.php [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    const CACHE_DIR = __DIR__ . '/../build/cache';
+
+    /**
+     * @var Client[]
+     */
+    private static $clients = [];
+
+    public static function setUpBeforeClass()
+    {
+        self::$clients = [
```
Comment (realityking) [inline tests/End2EndTestCase.php]: What's the reason for the eager creation? Intuitively I'd have implemented this the same as the CMA, it's easier to reason about.

File: tests/E2E/ErrorTest.php [hunk 6]
```diff
@@ -40,20 +40,23 @@ public function testAccessTokenInvalid()
...
+        $client = $this->getClient('cfexampleapi');
+
         $query = (new Query())
             ->where('name', 0);
 
         $client->getEntries($query);
     }
 
     /**
+     * @requires Contentful
```
Comment (realityking) [inline tests/E2E/ErrorTest.php]: Clever! Let me think about that one for a bit 😄

File: src/Client.php [hunk 7]
```diff
@@ -243,6 +243,14 @@ protected function getExceptionMap()
     }
 
     /**
+     * @return LoggerInterface
+     */
+    public function getLogger()
```
Comment (realityking) [inline src/Client.php]: This should definitely go into the changelog.

File: tests/End2EndTestCase.php [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+            'cfexampleapi_logger' => new Client('b4c0n73n7fu1', 'cfexampleapi', false, null, ['logger' => new ArrayLogger()]),
+            'cfexampleapi_tlh' => new Client('b4c0n73n7fu1', 'cfexampleapi', false, 'tlh'),
+            'cfexampleapi_invalid' => new Client('e5e8d4c5c122cf28fc1af3ff77d28bef78a3952957f15067bbc29f2f0dde0b50', 'cfexampleapi'),
+            '88dyiqcr7go8' => new Client('668efbfd9e398181166dec5df5a500aded96dbca2916646a3c7ec37082a7b756', '88dyiqcr7go8'),
+            '88dyiqcr7go8_preview' => new Client('81c469d7241ca02349388602dfc14107157063a6901c378a56e1835d688970bf', '88dyiqcr7go8', true),
+            'bc32cj3kyfet_preview' => new Client('8740056d546471e0640d189615470cc12ce2d3188332352ecfb53edac59c4963', 'bc32cj3kyfet', true),
+        ];
+    }
+
+    public function setUp()
```
Comment (realityking) [inline tests/End2EndTestCase.php]: This one I'm not a fan of in the parent class. That code will run before/after every single test method, whether it uses the cache or not.

File: src/Exception/ApiException.php (hunk 1 above)
Comment (dborsatto) [inline src/Exception/ApiException.php]: The coverage tools strips out all headers by default, so this raised a warning because that header is not available.

PSR-7 assumes that headers can have multiple values, and for some it _can_ be true (like set-cookie), but the vast majority is basically a key => value scenario. Because of this, `getHeaderLine` is simply a shortcut instead of doing something like
``` php
$requestIdHeader = $this->response->getHeader('X-Contentful-Request-Id');
$this->requestId = $requestIdHeader ? $requestIdHeader[0] : '';
```

`X-Contentful-Request-Id` will only ever be a single line, so there's really no problem using `getHeaderLine`, and on the plus side we don't have to make an explicit check ourselves.

File: tests/E2E/GzipEncodingTest.php (hunk 3 above)
Comment (dborsatto) [inline tests/E2E/GzipEncodingTest.php]: Yes, Guzzle lowercases everything, so this is just for consistency with routine use.

File: tests/End2EndTestCase.php (hunk 5 above)
Comment (dborsatto) [inline tests/End2EndTestCase.php]: The CMA client only needs a token, so that's why that approach works there. Here we need a bunch of different clients with very different requirements, so whereas the CMA gets away with `getReadOnlySpaceManager`/`getReadWriteSpaceManager`, here we'd need 9 different getter methods, or one with a huge switch/case, where the client would be instantiated on-demand and then saved.

But as the test suite is most often execute as a whole, it actually make sense to have the clients defined like this (instead of created/destroyed multiple times), and I could actually improve on this by checking that if the array is not empty, then returning early because clients are already set up.

File: tests/E2E/ErrorTest.php (hunk 6 above)
Comment (dborsatto) [inline tests/E2E/ErrorTest.php]: PHPUnit has this builtin with something like the PHP version or the existence of certain functions, but it was actually easy to extend. I find it nice and declarative.

File: tests/End2EndTestCase.php (hunk 8 above)
Comment (dborsatto) [inline tests/End2EndTestCase.php]: Yeah, I was trying to make something smarter than this, but in the end the performance impact is basically negligible...

Maybe I could implement this using an annotation like I did with `@requires`, something like `@uses-cache`?

File: tests/bootstrap-api-coverage.php (hunk 4 above)
Comment (dborsatto) [inline tests/bootstrap-api-coverage.php]: I'd like to avoid what I did in the CMA (with a single commit changing everything) but instead changing the header whenever the file is being changed...
This also needed updating, as it was saying 2015-2016, so it was two birds with one stone.

File: phpunit-api-coverage.xml (hunk 2 above)
Comment (dborsatto) [inline phpunit-api-coverage.xml]: I didn't actually think about it, to be honest.
Do you mean simply changing this to use `<env>` instead of `<const>` ([docs](https://phpunit.de/manual/current/en/appendixes.configuration.html#appendixes.configuration.php-ini-constants-variables)), or actually having environment variables set in whatever thing is running the coverage tool?

Comment (realityking) [inline phpunit-api-coverage.xml]: The former, using `<env>`.

My logic would be that environment variables were designed for these things. Having phpunit generate constants trips up static analysis, for example PhpStorm will claim that `SDK_TESTING_ENV` is an undefined constant.

File: src/Exception/ApiException.php (hunk 1 above)
Comment (realityking) [inline src/Exception/ApiException.php]: My concern were more around what happens if we have a bug and duplicate a header. (has happened before)

Of course with the current code we have the opposite problem that if we have a bug that the header isn't set if tails 🤔

File: tests/End2EndTestCase.php (hunk 8 above)
Comment (realityking) [inline tests/End2EndTestCase.php]: I think this is overkill. The whole cache thing is only used in a single class, why move the handling for it to the base class?

File: tests/bootstrap-api-coverage.php (hunk 4 above)
Comment (realityking) [inline tests/bootstrap-api-coverage.php]: Personally I'd rather see the big change then this. Otherwise this repo will be inconsistent for years...

File: tests/End2EndTestCase.php (hunk 5 above)
Comment (realityking) [inline tests/End2EndTestCase.php]: The big downside of this approach is that it breaks the isolation between test cases. This is most obvious in the `cfexampleapi_logger` case where the logger is shared across all tests using it. (It's only one right now)

Let me think about this one a bit too :)

File: phpunit-api-coverage.xml (hunk 2 above)
Comment (dborsatto) [inline phpunit-api-coverage.xml]: Yeah, Atom does the same. I'll change this 👍

File: tests/End2EndTestCase.php (hunk 8 above)
Comment (dborsatto) [inline tests/End2EndTestCase.php]: To be honest I don't remember why I started moving the cache management to the base class 🤔

File: src/Exception/ApiException.php (hunk 1 above)
Comment (dborsatto) [inline src/Exception/ApiException.php]: > My concern were more around what happens if we have a bug and duplicate a header. (has happened before)

What bug?

**Comment by dborsatto** [comment]
@realityking I've updated the way Client objects are created, please take a look.

File: src/Exception/ApiException.php (hunk 1 above)
Comment (realityking) [inline src/Exception/ApiException.php]: I mean if our API has a bug that a header gets sends twice. Similar things have happened to us.

I think neither implementation is 100% robust for corner cases. We need something like `getFirstHeaderLine()`.

File: tests/End2EndTestCase.php [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    /**
+     * @param string $key
+     *
+     * @return Client
+     */
+    protected function getClient($key)
+    {
+        $options = getenv('CONTENTFUL_CDA_SDK_TESTING_URL')
```
Comment (realityking) [inline tests/End2EndTestCase.php]: Stylistic but IMHO it's nicer to only call `getenv` once and then check/use the variable.

File: tests/End2EndTestCase.php [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+            ? ['uriOverride' => getenv('CONTENTFUL_CDA_SDK_TESTING_URL')]
+            : [];
+
+        switch ($key) {
```
Comment (realityking) [inline tests/End2EndTestCase.php]: This looks much better. :) Any particular reason you went for a switch statement instead of many methods?

One advantage of methods would be that we could parameterize them which would cut down on the number of different methods/cases we need and we could inject things like the logger or cache path at that point.

Though overall not a big difference.

File: src/Exception/ApiException.php (hunk 1 above)
Comment (dborsatto) [inline src/Exception/ApiException.php]: To be honest, making some weird workarounds for possible bugs that may not ever happen for a very specific implementation where the worst thing that could happen is to actually report the ID twice (should such a bug ever arise), well, it seems a bit silly to me...

File: tests/E2E/ErrorTest.php (hunk 6 above)
Comment (realityking) [inline tests/E2E/ErrorTest.php]: While I'm not the biggest annotations fan, this seems fine.

What I don't think is good is the name. All E2E tests require Contentful, what makes these different?

File: tests/End2EndTestCase.php [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+        foreach (['class', 'method'] as $depth) {
+            if (empty($annotations[$depth]['requires'])) {
+                continue;
+            }
+
+            $requires = array_flip($annotations[$depth]['requires']);
+
+            if (isset($requires['Contentful'])) {
+                return $this->markTestSkipped('Tests against Preview API are not enabled with this configuration.');
```
Comment (realityking) [inline tests/End2EndTestCase.php]: I think this is not 100% accurate, some of the skipped tests don't use the Preview API. (e.g. the URI override one)

File: tests/E2E/ErrorTest.php (hunk 6 above)
Comment (dborsatto) [inline tests/E2E/ErrorTest.php]: Because they require to speak _directly_ to Contentful, and not a proxy of some kind. I can change this to `@requires Contentful no-proxy`, if you think it looks better

File: src/Exception/ApiException.php (hunk 1 above)
Comment (realityking) [inline src/Exception/ApiException.php]: 🤷‍♂️ Once we go down the path of hardening this might as well do it right

I don't feel too strongly about this. We use `getHeader()` all over the place (check the RateLimitException classes in both SDKs for others), it might be nice to just roll a helper function somewhere that takes care of this.

File: tests/End2EndTestCase.php (hunk 11 above)
Comment (dborsatto) [inline tests/End2EndTestCase.php]: Yeah, I wrote that message before the URI override one was added to the list, I'll rephrase this

File: tests/E2E/ErrorTest.php (hunk 6 above)
Comment (realityking) [inline tests/E2E/ErrorTest.php]: Oh wow, I didn't guess that meaning at all.

How about something like `@requires No coverage proxy`? That makes it clear we're not talking about normal HTTP proxies (which should work)

File: src/Exception/ApiException.php (hunk 1 above)
Comment (dborsatto) [inline src/Exception/ApiException.php]: Honestly I'd rather patch the coverage tool to let this header pass through (which would make a lot of sense anyway) than going down the path of fixing possible header issues (and reverting this change), what do you think?

Comment (realityking) [inline src/Exception/ApiException.php]: Patching the coverage is probably a good idea either way. Again I don't feel to strongly whether this line looks one way or another but if we change it, we should change it everywhere.

File: tests/E2E/ErrorTest.php (hunk 6 above)
Comment (dborsatto) [inline tests/E2E/ErrorTest.php]: Standard `@requires` annotations have a format `@requires SUBJECT ATTRIBUTE`, such as `@requires PHP 7.0` or `@requires extension ext-soap`. I suggest we use a similar structure for consistency, maybe `@requires API no-proxy`

File: tests/End2EndTestCase.php (hunk 9 above)
Comment (dborsatto) [inline tests/End2EndTestCase.php]: No real difference but sure 👍

File: tests/E2E/ErrorTest.php (hunk 6 above)
Comment (realityking) [inline tests/E2E/ErrorTest.php]: Fair enough. Just trying to make it clear we're not talking about normal HTTP proxies. (It'd be fun to rewrite the coverage tool being an actual proxy!)

`@requires API no-coverage-proxy`?

File: tests/End2EndTestCase.php (hunk 10 above)
Comment (dborsatto) [inline tests/End2EndTestCase.php]: Mostly because not all methods would have meaningful names.
For instance `cfexampleapi` and `88dyiqcr7go8` (and their respective `_preview` versions) are for all intents and purposes the same, but the things they need for testing are in two different spaces.

Some of them would be hard to describe using a name, so we might as well not do that and just use a key...

File: tests/E2E/ErrorTest.php (hunk 6 above)
Comment (dborsatto) [inline tests/E2E/ErrorTest.php]: Sounds good 👍

File: src/Exception/ApiException.php (hunk 1 above)
Comment (dborsatto) [inline src/Exception/ApiException.php]: Because doing that would be out of scope, I'll patch the tool and postpone this discussion 🙂

File: tests/End2EndTestCase.php (hunk 10 above)
Comment (realityking) [inline tests/End2EndTestCase.php]: Now that you mention it, that might actually be a problem. I can't tell anymore which space is the one used for rate limit tests for example.

But let's keep it like this for now, can always refactor it later.

File: src/Client.php (hunk 7 above)
Comment (realityking) [inline src/Client.php]: Based on the discussion downstream. Do you think this has merits on its own or only for testing? If it's the latter I suggest we refactor `getClient()` a tiny bit.

File: tests/End2EndTestCase.php (hunk 10 above)
Comment (dborsatto) [inline tests/End2EndTestCase.php]: Well, same as before, you can look at the `ErrorTest.php` and see the space being used, it doesn't really change that 🙂

File: src/Client.php (hunk 7 above)
Comment (dborsatto) [inline src/Client.php]: I feel that's actually something worth doing either way.
For instance, this makes what I wanted to do in https://github.com/contentful/contentful.php/pull/162 much easier, as you don't need to keep the logger around because you just can retrieve it from the client.

The fact that it was added to cover a specific need here doesn't mean it'd not be useful overall, on the contrary...

Comment (realityking) [inline src/Client.php]: 👍 

Though I disagree that this would be any help for #162 - the default `NullLogger` won't do you any good.

Comment (dborsatto) [inline src/Client.php]: Yeah, but it would be a matter of ease of use.
Maybe you're setting up the client somewhere, and using it somewhere else. This way you don't have to keep yourself the logger around, you can simply rely on the client to store it for you...

**Comment by dborsatto** [comment]
I've updated:
- Annotations (now `@requires API no-coverage-proxy`)
- Reverted `getHeaderLine` in `ApiException`
- Changed the skipped test error message
- Stored the result of `getenv` in a local variable

I think I've address all concerns, am I missing something?

File: phpunit.xml.dist [hunk 12]
```diff
@@ -28,4 +27,5 @@
             <directory suffix=".php">src</directory>
         </whitelist>
     </filter>
+
```
Comment (realityking) [inline phpunit.xml.dist]: This looks like an accident ;)

Comment (dborsatto) [inline phpunit.xml.dist]: I had added the env config to that file as well, but later I've removed it, that's why there a leftover space there. I'll remove it 👍
